lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,125 @@
+# it.
+#
+# That was v0 of the auth cookie until Nov 2021 when we merged a change that
+# required us to store additional metadata with the cookie so we could get more
+# information about current user early in the request lifecycle before we
+# performed database lookup. We also started encrypting and signing the cookie
+# to prevent tampering and obfuscate user information that we include in the
+# cookie. This is v1 of our auth cookie and we still use it to this date.
+#
+# We still accept v0 of the auth cookie to keep users logged in, but upon
+# cookie rotation (which happen every 10 minutes) they'll be switched over to
+# the v1 format.
+#
+# We'll drop support for v0 after Discourse 2.9 is released.
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        db_name = RailsMultisite::ConnectionManagement.current_db
+        encryption_key = Rails.application.key_generator.generate_key(
+          "discourse-auth-cookie-encryption-#{db_name}",
+          32
+        )
+        sign_key = Rails.application.key_generator.generate_key(
+          "discourse-auth-cookie-sign-#{db_name}",
+          32
+        )
+        ActiveSupport::MessageEncryptor.new(
+          encryption_key,
+          sign_key,
+          cipher: "aes-256-cbc",
+          digest: "SHA256"
+        )
+      end
+    end
+  end
+
+  TOKEN_SIZE = 32
+
+  TOKEN_KEY = "token"
+  USER_ID_KEY = "id"
+  USER_TRUST_LEVEL_KEY = "tl"
+  ISSUED_AT_KEY = "issued_at"
+  private_constant *%i[
+    TOKEN_KEY
+    USER_ID_KEY
+    USER_TRUST_LEVEL_KEY
+    ISSUED_AT_KEY
+  ]
+
+  attr_reader :token, :user_id, :trust_level, :issued_at
+
+  def self.parse(raw_cookie)
+    # v0 of the cookie was simply the auth token itself. we need this for
+    # backward compatibility so we don't wipe out existing sessions.
+    # TODO: drop this line after the 2.9 release
+    return new(token: raw_cookie) if raw_cookie.size == TOKEN_SIZE
+
+    json = Encryptor.new.decrypt_and_verify(raw_cookie)
+    data = JSON.parse(json)
+
+    token = data[TOKEN_KEY]
+    user_id = data[USER_ID_KEY]
+    trust_level = data[USER_TRUST_LEVEL_KEY]
+    issued_at = data[ISSUED_AT_KEY]
+
+    new(
+      token: token,
+      user_id: user_id,
+      trust_level: trust_level,
+      issued_at: issued_at
+    )
+  end
+
+  def initialize(token:, user_id: nil, trust_level: nil, issued_at: nil)
+    @token = token
+    @user_id = user_id.to_i if user_id
+    @trust_level = trust_level.to_i if trust_level
+    @issued_at = issued_at.to_i if issued_at
+    validate!
+  end
+
+  def serialize
```

Comment:
I was just reading through the docs for [MessageEncryptor](https://api.rubyonrails.org/v5.2.3/classes/ActiveSupport/MessageEncryptor.html), and it looks like it actually has built-in support for serialization. By default it uses Marshall (and therefore, this is what we use for the `session` cookie).

By creating JSON here, we're 'double serializing' the value. We're serializing to JSON, and then the MessageEncryptor is serializing that string object using Marshal. This adds some extra unnecessary characters to the result:

```
[1] pry(main)> {some: "hash"}.to_json
=> "{\"some\":\"hash\"}"
[2] pry(main)> Marshal.dump(_)
=> "\x04\bI\"\x14{\"some\":\"hash\"}\x06:\x06ET"
```

Instead, we could pass a hash straight into `encrypt_and_sign`. This would be encoded using Marshall, and automatically decoded when we later `decrypt_and_verify`:

```
[4] pry(main)> {some: "hash"}
=> {:some=>"hash"}
[5] pry(main)> Marshal.dump(_)
=> "\x04\b{\x06:\tsomeI\"\thash\x06:\x06ET"
[6] pry(main)> Marshal.load(_)
=> {:some=>"hash"}
```

(If we really want JSON, rather than Marshall, we could pass `serializer: JSON` to `MessageEncryptor`. But I think it makes sense to keep `Marshal` here, since it matches the existing session cookie.)

Reply:
Excellent catch, thanks! If we can rely on `ActionDispatch::Request` to do the encryption for us, I think this problem will go away.